From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wosnxn123 <wosnxn123@users.noreply.github.com>
Date: Thu, 7 Aug 2026 12:00:00 +0800
Subject: [PATCH] MiniTweaks mob fire and explosion rules

Ported from Lophine 0125 (MiniTweaks, GPL-3.0, Bacteriawa).
Gated behind the four no*/disable* flags.

diff --git a/net/minecraft/world/entity/monster/Creeper.java b/net/minecraft/world/entity/monster/Creeper.java
index 4a90294f8d8447bb2f2cef7572119274a1a2c62e..3b8a0c17c2ba8d00e177aa9a6d15e04b38f9c5a9 100644
--- a/net/minecraft/world/entity/monster/Creeper.java
+++ b/net/minecraft/world/entity/monster/Creeper.java
@@ -250,7 +250,8 @@ public class Creeper extends Monster {
             if (!event.isCancelled()) {
             // CraftBukkit end
             this.dead = true;
-            level.explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.MOB); // CraftBukkit // Paper - fix DamageSource API (revert to vanilla, no, just no, don't change this)
+            Level.ExplosionInteraction explosionInteraction = io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaLikeExperience.noCreeperBlockBreaking ? Level.ExplosionInteraction.NONE : Level.ExplosionInteraction.MOB; // Lophine - MiniTweaks noCreeperBlockBreaking
+            level.explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), explosionInteraction); // CraftBukkit // Paper - fix DamageSource API (revert to vanilla, no, just no, don't change this)
             this.spawnLingeringCloud();
             this.triggerOnDeathMobEffects(level, Entity.RemovalReason.KILLED);
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit - add Bukkit remove cause
diff --git a/net/minecraft/world/entity/projectile/hurtingprojectile/LargeFireball.java b/net/minecraft/world/entity/projectile/hurtingprojectile/LargeFireball.java
index 8195ab561d12abb33f4359a68b3c1579ce284242..9984244d549ad08a6053fcd74344667f1e1bc4f1 100644
--- a/net/minecraft/world/entity/projectile/hurtingprojectile/LargeFireball.java
+++ b/net/minecraft/world/entity/projectile/hurtingprojectile/LargeFireball.java
@@ -38,7 +38,9 @@ public class LargeFireball extends Fireball {
             // CraftBukkit start - fire ExplosionPrimeEvent
             org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent((org.bukkit.entity.Explosive) this.getBukkitEntity());
             if (event.callEvent()) {
-                this.level().explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.MOB);
+                Level.ExplosionInteraction explosionInteraction = io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaLikeExperience.noGhastBlockBreaking ? Level.ExplosionInteraction.NONE : Level.ExplosionInteraction.MOB; // Lophine - MiniTweaks noGhastBlockBreaking
+                boolean createFire = event.getFire() && !io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaLikeExperience.disableGhastFire; // Lophine - MiniTweaks disableGhastFire
+                this.level().explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), createFire, explosionInteraction);
             }
             // CraftBukkit end - fire ExplosionPrimeEvent
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
diff --git a/net/minecraft/world/entity/projectile/hurtingprojectile/SmallFireball.java b/net/minecraft/world/entity/projectile/hurtingprojectile/SmallFireball.java
index ea2ed37063a5f62ad2ade39e9f39f2c2ba6f1772..64eb290ffd73305054a805a1d698170c18742902 100644
--- a/net/minecraft/world/entity/projectile/hurtingprojectile/SmallFireball.java
+++ b/net/minecraft/world/entity/projectile/hurtingprojectile/SmallFireball.java
@@ -64,7 +64,7 @@ public class SmallFireball extends Fireball {
         super.onHitBlock(hitResult);
         if (this.level() instanceof ServerLevel serverLevel) {
             Entity owner = this.getOwner();
-            if (this.isIncendiary) { // CraftBukkit
+            if (this.isIncendiary && (!(owner instanceof Mob) || !io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaLikeExperience.disableBlazeFire)) { // CraftBukkit // Lophine - MiniTweaks disableBlazeFire
                 BlockPos pos = hitResult.getBlockPos().relative(hitResult.getDirection());
                 if (this.level().isEmptyBlock(pos) && !org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(this.level(), pos, this).isCancelled()) { // CraftBukkit
                     this.level().setBlockAndUpdate(pos, BaseFireBlock.getState(this.level(), pos));
